### ISSUE TITLE
Selectively sync primary index to DataStore using query operation instead of scan operation

### ISSUE THREAD

**Issue by htc007** [aws-amplify/amplify-js#7778]
**Is your feature request related to a problem? Please describe.**
According to the docs for Amplify DataStore, you can selectively sync a subset of your data by specifying a sync expression.

If I construct a sync expression based on the partition key of the primary index (not a GSI), will the base sync perform a scan operation or a query operation? This is not clear in the docs.

If the base table contains a million items, but I only need to sync 100 specified items, a scan operation will be really inefficient and costly.

Can you clarify whether specifying the partition key of the primary index in the sync expression will perform a scan operation or a query operation.

**Describe the solution you'd like**
Perform a query operation instead of a scan operation when syncing data from the primary index.

**Describe alternatives you've considered**
I understand that it's possible to define a sync expression that returns the partition key of a GSI (instead of the primary index) and this will perform a query operation. However, this requires me to create a GSI with a duplicate set of data just for this purpose.

**Additional context**
Add any other context, an idea of the implementation or screenshots about the feature request here.

**Comment by iartemiev** [comment]
Hey @htc007, that is correct. You cannot currently perform a query on a model's PK, only on a GSI. 

Given that the PK has to be unique, how would you construct a query to retrieve 100 items by PK? It seems you would need to have a GSI for that either way unless I'm missing something.

**Comment by htc007** [comment]
My example's not great. What I meant to say was if the table had a PK as well as a sort key. In this case, the PK wouldn't be unique. For example, assume I have a single-table design where Artist is the PK and SongTitle is the SK. My objective is to sync all the song titles of a particular artist. How would I achieve this with selective sync?

If I were to specify the artist in the sync expression without creating a GSI, this would scan the entire base table just to sync that artists' songs right? This seems overkill if the table contains many artists, albums and other related data.

I'm not sure if I misunderstood the docs, but if I were to use a partition key value to query the primary index using the GraphQL API, this would perform a query operation instead of a scan operation and return all items with that partition key value (much like querying DynamoDB directly).

However, this doesn't seem to apply to DataStore's selective sync.

**Comment by iartemiev** [comment]
You wouldn't be able to do that with DataStore since it automatically generates an `id` field as the PK and populates it with a UUID.

**Comment by htc007** [comment]
Based on our testing, AppSync automatically generates an id field and populates it with a UUID regardless of whether we enable DataStore.

Here's where I'm a little confused:

When we use the @key directive to define our model, we set our own partition key and sort key using the fields argument. If we subsequently enable DataStore on our API, how does the id field become the partition key? Do you mean DataStore itself and the delta sync table have their own partition key, this being the id field?

Can you clarify what goes on behind the scenes between the primary index and the delta sync table if the partition key is different?

Why doesn't selectively syncing data against a GSI suffer from the same problems as selectively syncing against the primary index?

**Comment by iartemiev** [comment]
> If we subsequently enable DataStore on our API, how does the id field become the partition key?

The `id` field won't become the PK automatically. DataStore simply will not work with a custom PK. See the callout at the top of [this section ](https://docs.amplify.aws/lib/datastore/relational/q/platform/js) of the docs.

There's no way to modify the PK on an existing DynamoDB table. You would need to remove the table and re-create it with the default PK, i.e., `id`.

> Why doesn't selectively syncing data against a GSI suffer from the same problems as selectively syncing against the primary index?

The limitation here is not with selective sync per se, but the fact that DataStore currently requires every model's PK to be the `id` field with a unique value. Since the `id` has to be unique, there doesn't seem to be much benefit in allowing customers to specify a sync expression that maps to a query on the PK, as it would return at most a single record.

**Comment by htc007** [comment]
I would suggest putting this callout in a more prominent place in the DataStore docs. Currently, it's in the relational models section. I did see it before, but I thought it only applied to models with relationships.

Oddly, we did occasionally manage to sync data with DataStore even though our PK was not the `id` field, though the PK value was also a UUID.

Is there a roadmap for supporting DynamoDB tables with a custom PK and SK? I can imagine many (if not most) people define their own PK and SK as recommended by the docs for the `@key` directive itself.

It's a bit disheartening for those who spend time structuring their tables based on their access patterns only to find out that it doesn't work DataStore.
